pkg/kn/commands/trigger/update.go

```diff
@@ -82,12 +82,7 @@ func NewTriggerUpdateCommand(p *commands.KnParams) *cobra.Command {
 					return fmt.Errorf(
 						"cannot update trigger '%s' because %s", name, err)
 				}
-				for k, v := range updated {
-					b.AddFilter(k, v)
-				}
-				for _, k := range removed {
-					b.RemoveFilter(k)
-				}
+				b.Filters(updated).RemoveFilters(removed)
```

Comment:
Sorry, my last nit-pick (promised :). What do you think about:

```golang
b.Filters(maputil.RemoveKeysFromMap(updated, removed))
```

and a `maputil.RemoveFrom()` (but maybe in an already existing util package):

```golang
func RemoveKeysFromMap(original map[string]string, toRemove []string) map[string]string {
     ....
}
```

That way this method could be re-used for other use cases and we don't have to add a `Remove` method to a builder.

wdyt ?

Reply:
never mind.
The problem of your suggestion is that I need to get the original value of filters map to remove. I think we want to hide the detail of trigger `Spec` by Builder. Except I add a method to Builder `GetFilters()`, I have to manage the trigger `Spec` structure to get the filter. The code may like below:
```
var attributes *map[string]string
attributes = nil
if b.Build().Spec.Filter != nil {
    attributes = b.Build().Spec.Filter.Attributes
}
b.Filter(maputil.AddAndRemoveFromMap(attributes, updated,removed))
```

and
```
func AddAndRemoveFromMap(original *map[string]string, toAdd map[string]string, toRemove []string) map[string]string {
......
}
```

If I add a `GetFilters()` to Builder, it may look like
```
b.Filter(maputil.AddAndRemoveFromMap(b.GetFilters(), updated,removed))
```
But it violates the rule that builder is to build something.

Do I make sense ?

Reply:
My suggestion:

```golang
// Extract filters directly from original trigger (not builder). Return nil if none.
existing := extractFilters(trigger)

// Use chained util method, so that they can bey reused in different contexts:
// * type StringMap map[string]string
// * func StringMap(map [string]string) StringMap : Just cast or copy to the new type (then call it "NewStringMap")
// * func (m SringMap) Merge(toMerge map[string]string) StringMap 
//    Keys from the toMerge map override those from the wrapped map
// * func (m map[string]string) RemovKeys(toRemove string[]) for removing the keys of a map. If this no
b.Filters(maputil.StringMap(existing).Merge(updated).Remove(removed))
```

This would have the benefit:

* Builder is still immutable and write-only
* Reusable string map manipulation functions which can be easily used in different contexts
* Easy to use API (maybe a bit harder to implement but obey one of the principles of good API design: "It should be as easy to use as possible, even when its might harder to implement that parts that are hidden from the user". 

Instead of introducing a "StringMap" you could equally cast to the type "StringMap" via `existing.(StringMap).Merge(updated).Remove(toRemove)`

Does this make sense ?

Reply:
Thanks. I made a few update to your suggestion.

I used builder as parameter in extractFilters, just to make sure no dependencies to `knative.dev/eventing/pkg/apis/eventing/v1alpha1` in the update command. Here is the definition of `extractFilters`
```
func extractFilters(builder *client_v1alpha1.TriggerBuilder) util.StringMap
```

Reply:
> I used builder as parameter in `extractFilters`, not trigger. I don't like to introduce a dependency to `knative.dev/eventing/pkg/apis/eventing/v1alpha1` in the update command. If I use trigger as parameter, I have to add this dependency. Here is the definition of `extractFilters`

Don't we already have the trigger available in this method ? with `extractFilters()` I mean just a function within `update.go` itself (just holding the loop that you've mentioned). I don't think we should put that into the builder.

I refer to this line: https://github.com/knative/client/blob/5c2ac9682343b141af049aee8dd084104fd14918/pkg/kn/commands/trigger/update.go#L69

Reply:
Ah, you mean a goimport. Good point, but we have the implicit dependency anyway (e.g. via the constructor arg of the builder or the return value). Alternatively you could add the function to client_v1alpha1 package, like `client_v1alpha1.ExtractFilters(trigger)`


It's better than diving into a builder. E.g. in generally when you call `Build()` on a Builder this is considered often to be a final action, so you can't call Build() a second time on the same builder (because it performs some more persisten action). So I would prefer to avoid that in favor of the utility in the client package.

Reply:
OK. I added a line to goimport in `update.go`, and used trigger as input parameter in the latest version. Please review. Thank you.